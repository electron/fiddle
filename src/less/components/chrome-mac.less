.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 24px;
    font-size: 1.4rem;
    line-height: 15px;

    small,
    p,
    span {
      user-select: none;
    }
  }

  .big_sur_or_later {
    height: 30px !important;
  }
}
