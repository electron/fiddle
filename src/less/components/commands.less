@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-3;
  -webkit-app-region: drag;
}

.disabled-menu-tooltip {
  display: flex;
  width: 100%;

  .bp3-popover-target {
    width: 100%;
  }
}

.commands.is-mac {
  padding:
    10px
    calc(100vw - (env(titlebar-area-x, 0) + env(titlebar-area-width, 0)))
    10px
    env(titlebar-area-x, 0);
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding:
    10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
    white-space: nowrap;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  fieldset {
    border: 0;
    padding: 0;
    margin: 0;
    min-width: 0;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      .bp3-button-group {
        margin-left: 5px;
      }
    }
  }

  .bp3-control-group {
    margin-right: 0.5rem;
  }

  .address-bar {
    width: 420px;
    transition: width 0.4s;

    input,
    span {
      -webkit-app-region: no-drag;
    }

    &.empty {
      width: 300px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }

  & > .title {
    align-self: stretch;
    flex-shrink: 0;

    flex-flow: column nowrap;
    justify-content: center;
    user-select: none;

    margin: 0 10px;
  }
}
