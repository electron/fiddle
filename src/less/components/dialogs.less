@import "../variables.less";

.dialogs {
  z-index: 10;
  position: fixed;
  height: 100vh;
  width: 100vw;
  pointer-events: none;

  * {
    pointer-events: all;
  }
}

.dialog {
  background-color: @background-3;
  border: @border;
  color: @text-color-1;
  border-radius: 6px;
  padding: 10px;
  z-index: 100;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  animation: slide-up 0.4s ease;

  p {
    margin-bottom: 1.25rem;
    font-size: 1.25rem;
  }

  code {
    font-size: 12px;
    background: rgba(255, 255, 255, 0.1);
    padding: 2px;
    line-height: 12px;
  }

  &-buttons {
    margin-top: 1.25rem;
  }

  &.centered {
    position: absolute;
    width: 300px;
    left: calc(50% - (300px / 2));
    top: 100px;
  }

  &.arrow::after,
  &.arrow::before {
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &.arrow::after {
    border-color: rgba(136, 183, 213, 0);
    border-width: 9px;
  }

  &.arrow::before {
    border-color: rgba(194, 225, 245, 0);
    border-width: 10px;
  }

  &.arrow-top::after,
  &.arrow-top::before {
    bottom: 100%;
    left: 50%;
  }

  &.arrow-top::after {
    border-bottom-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-top::before {
    border-bottom-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-bottom::after,
  &.arrow-bottom::before {
    top: 100%;
    left: 50%;
  }

  &.arrow-bottom::after {
    border-top-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-bottom::before {
    border-top-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-left::after,
  &.arrow-left::before {
    right: 100%;
    top: 20px;
  }

  &.arrow-left::after {
    border-right-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-left::before {
    border-right-color: @border-color-1;
    margin-top: -10px;
  }

  &.arrow-right::after,
  &.arrow-right::before {
    left: 100%;
    top: 20px;
  }

  &.arrow-right::after {
    border-left-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-right::before {
    border-left-color: @border-color-1;
    margin-top: -10px;
  }

  button {
    color: #1e2527;
    float: right;
    margin-left: 9px;
    margin-bottom: 0;
  }

  a {
    cursor: pointer;
  }

  h4 {
    margin-bottom: 0;
  }

  p:last-of-type {
    margin-bottom: 0;
  }

  input {
    color: rgba(159, 234, 249, 0.5);
    height: 28px;
    margin-top: 10px;
    background-color: @background-1;
    border: 1px solid @border-color-2;
    outline: 0 solid transparent;

    &:focus {
      border-color: @foreground-2;
    }
  }
}

.dialogDrop {
  position: fixed;
  z-index: 100;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: black;
  transition: opacity 0.4s ease-in-out;
  pointer-events: all;
  opacity: 0.5;
}

.dialog-add-version {
  .bp3-file-input {
    width: 100%;
  }

  .bp3-callout {
    margin-top: 1rem;
  }
}

select {
  font-size: 12px;
  height: 28px;
  background-color: white;
  line-height: 15px;
}

@keyframes slide-up {
  0% {
    opacity: 0;
    transform: translateY(20px);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
