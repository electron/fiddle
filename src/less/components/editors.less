.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 5px 0 5px;

    .mosaic-controls {
      transform: scale(0.75);
      transform-origin: right;
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

// TODO: support new file
// update if list of Editor ID changes
@editor-ids: main\.js, renderer\.js, index\.html, preload\.js, styles\.css;

// takes each editor ID and increase its z-index if the parent Mosaic root
// has focused__id class for that specific id.
each(@editor-ids, {
  .focused__@{value}.mosaic .mosaic-window.@{value} {
    z-index: 2;
  }
});

.mosaic-window {
  z-index: 1;
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
