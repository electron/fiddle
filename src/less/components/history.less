.revision-list {
  max-height: 400px;
  overflow-y: auto;
}

.revision-list ul {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.revision-item {
  padding: 10px;
  margin-bottom: 5px;
  border-radius: 3px;
  cursor: pointer;
  transition: background-color 0.2s ease;
  border-left: 3px solid #106ba3;
}

.revision-item:hover {
  background-color: rgba(167, 182, 194, 0.3);
}

.revision-content {
  display: flex;
  flex-direction: column;
}

.revision-icon {
  margin-right: 8px;
}

.sha-label {
  font-size: 12px;
  color: #738694;
  margin-left: 10px;
  font-family: monospace;
}

.revision-details {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 5px;
}

.revision-date {
  color: #738694;
  font-size: 12px;
}

.revision-changes {
  display: flex;
  gap: 5px;
}

.history-loading {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
}
