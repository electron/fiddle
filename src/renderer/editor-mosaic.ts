import { action, computed, makeObservable, observable, reaction } from 'mobx';
import * as MonacoType from 'monaco-editor';
import { MosaicDirection, MosaicNode, getLeaves } from 'react-mosaic-component';

import { EditorId, EditorValues } from '../interfaces';
import {
  compareEditors,
  getEmptyContent,
  isSupportedFile,
  monacoLanguage,
} from '../utils/editor-utils';

export type Editor = MonacoType.editor.IStandaloneCodeEditor;

export enum EditorPresence {
  /** The file is known to us but we've chosen not to show it, either
      because the content was boring or because hide() was called.
      Its contents are cached offscreen. */
  Hidden,

  /** Space has been allocated for this file in the mosaic but the
      monaco editor has not mounted in React yet. This is an interim
      state before the editor is Visible. */
  Pending,

  /** The file is visible in one of the mosaic's monaco editors */
  Visible,
}

interface EditorBackup {
  model: MonacoType.editor.ITextModel;
  viewState?: MonacoType.editor.ICodeEditorViewState | null;
}

export class EditorMosaic {
  public isEdited = false;
  public focusedFile: EditorId | null = null;

  public get files() {
    const files = new Map<EditorId, EditorPresence>();

    const { backups, editors, mosaic } = this;
    for (const id of backups.keys()) files.set(id, EditorPresence.Hidden);
    for (const id of getLeaves(mosaic)) files.set(id, EditorPresence.Pending);
    for (const id of editors.keys()) files.set(id, EditorPresence.Visible);

    return files;
  }

  public get numVisible() {
    return getLeaves(this.mosaic).length;
  }

  // You probably want EditorMosaic.files instead.
  // This is only public because components/editors.tsx needs it
  public mosaic: MosaicNode<EditorId> | null = null;

  private readonly backups = new Map<EditorId, EditorBackup>();
  private readonly editors = new Map<EditorId, Editor>();

  constructor() {
    makeObservable<
      EditorMosaic,
      'backups' | 'editors' | 'addFile' | 'setVisible' | 'setEditorFromBackup'
    >(this, {
      isEdited: observable,
      focusedFile: observable,
      files: computed,
      numVisible: computed,
      mosaic: observable,
      backups: observable,
      editors: observable,
      setFocusedFile: action,
      resetLayout: action,
      set: action,
      addFile: action,
      show: action,
      setVisible: action,
      toggle: action,
      hide: action,
      remove: action,
      addEditor: action,
      setEditorFromBackup: action,
      addNewFile: action,
    });

    // whenever the mosaics are changed,
    // update the editor layout
    reaction(
      () => this.mosaic,
      () => this.layout(),
    );

    // whenever isEdited is set, stop or start listening to edits again.
    reaction(
      () => this.isEdited,
      () => {
        if (this.isEdited) {
          this.ignoreAllEdits();
        } else {
          this.observeAllEdits();
        }
      },
    );
  }

  /** File is visible, focus file content */
  /** File is hidden, show the file and focus the file content */
  public setFocusedFile(id: EditorId) {
    this.focusedFile = id;
    if (this.files.get(this.focusedFile) === EditorPresence.Hidden) {
      this.show(this.focusedFile);
    }
    this.editors.get(id)?.focus();
  }

  /** Reset the layout to the initial layout we had when set() was called */
  resetLayout = () => {
    this.set(this.values());
  };

  /// set / add / get the files in the model

  /** Set the contents of the mosaic */
  public set(valuesIn: EditorValues) {
    // set() clears out the previous Fiddle, so clear our previous state
    // except for this.editors -- we recycle editors below in setFile()
    this.backups.clear();
    this.mosaic = null;

    // add the files to the mosaic, recycling existing editors when possible.
    const values = new Map(Object.entries(valuesIn)) as Map<EditorId, string>;
    for (const [id, value] of values) this.addFile(id, value);
    for (const id of this.editors.keys()) {
      if (!values.has(id)) this.editors.delete(id);
    }

    this.isEdited = false;
  }

  /** Add a file. If we already have a file with that name, replace it. */
  private addFile(id: EditorId, value: string) {
    if (!isSupportedFile(id))
      throw new Error(`Cannot add file "${id}": Must be .js, .html, or .css`);

    // create a monaco model with the file's contents
    const { monaco } = window.ElectronFiddle;
    const language = monacoLanguage(id);
    const model = monaco.editor.createModel(value, language);
    model.updateOptions({ tabSize: 2 });

    // if we have an editor available, use the monaco model now.
    // otherwise, save the file in `this.backups` for future use.
    const backup: EditorBackup = { model };
    const editor = this.editors.get(id);
    if (editor) {
      this.setEditorFromBackup(editor, backup);
    } else {
      this.backups.set(id, backup);
    }

    // only show the file if it has nontrivial content
    if (value.length && value !== getEmptyContent(id)) {
      this.show(id);
    } else {
      this.hide(id);
    }
  }

  /// show or hide files in the view

  /** Show the specified file's editor */
  public show(id: EditorId) {
    this.setVisible([...getLeaves(this.mosaic), id]);
  }

  private setVisible(visible: EditorId[]) {
    // Sort the files and remove duplicates
    visible = [...new Set(visible)].sort(compareEditors);

    // Decide what layout would be good for this set of files
    const mosaic = EditorMosaic.createMosaic(visible);

    // Use that new layout. Note: if there are new files in `visible`,
    // setting `this.mosaic` here is what triggers new Monaco editors to
    // be created for them in React, via components/editors.tsx's use of
    // this.mosaic in its render() function. After they mount,
    // editors.tsx will call addEditor() to tell us about them. The same
    // holds true for removing editors; setting this.mosaic is what will
    // trigger their removal from React.
    this.mosaic = mosaic;
  }

  private static createMosaic(
    input: EditorId[],
    direction: MosaicDirection = 'row',
  ): MosaicNode<EditorId> {
    // Return single editor or undefined.
    if (input.length < 2) return input[0];

    // This cuts out the first half of input. Input becomes the second half.
    const secondHalf = [...input];
    const firstHalf = secondHalf.splice(0, Math.floor(secondHalf.length / 2));

    return {
      direction,
      first: EditorMosaic.createMosaic(firstHalf, 'column'),
      second: EditorMosaic.createMosaic(secondHalf, 'column'),
    };
  }

  /** Helper to toggle visibility of the specified file's editor */
  public toggle(id: EditorId) {
    if (this.files.get(id) === EditorPresence.Hidden) {
      this.show(id);
    } else {
      this.hide(id);
    }
  }

  /** Hide the specified file's editor */
  public hide(id: EditorId) {
    const editor = this.editors.get(id);
    if (editor) {
      this.editors.delete(id);
      this.backups.set(id, {
        model: editor.getModel()!,
        viewState: editor.saveViewState(),
      });
    }

    this.setVisible(getLeaves(this.mosaic).filter((v) => v !== id));
  }

  /** Remove the specified file and its editor */
  public remove(id: EditorId) {
    this.editors.delete(id);
    this.backups.delete(id);
    this.setVisible(getLeaves(this.mosaic).filter((v) => v !== id));
  }

  /** Wire up a newly-mounted Monaco editor */
  public addEditor(id: EditorId, editor: Editor) {
    const backup = this.backups.get(id);
    if (!backup) throw new Error(`added Editor for unexpected file "${id}"`);

    this.backups.delete(id);
    this.editors.set(id, editor);
    this.setEditorFromBackup(editor, backup);
  }

  /** Populate a MonacoEditor with the file's contents */
  private setEditorFromBackup(editor: Editor, backup: EditorBackup) {
    this.ignoreEdits(editor); // pause this so that isEdited doesn't get set
    if (backup.viewState) editor.restoreViewState(backup.viewState);
    editor.setModel(backup.model);
    this.observeEdits(editor); // resume
  }

  /** Add a new file to the mosaic */
  public addNewFile(id: EditorId, value: string = getEmptyContent(id)) {
    if (this.files.has(id))
      throw new Error(`Cannot add file "${id}": File already exists`);

    this.addFile(id, value);
  }

  /** Get the contents of a single file. */
  public value(id: EditorId): string {
    const { backups, editors } = this;
    return (
      editors.get(id)?.getValue() || backups.get(id)?.model.getValue() || ''
    );
  }

  /** Get the contents of all files. */
  public values(): EditorValues {
    return Object.fromEntries(
      [...this.files].map(([id]) => [id, this.value(id)]),
    );
  }

  /// misc utilities

  private layoutDebounce: ReturnType<typeof setTimeout> | undefined;

  public layout = () => {
    const DEBOUNCE_MSEC = 50;
    if (!this.layoutDebounce) {
      this.layoutDebounce = setTimeout(() => {
        for (const editor of this.editors.values()) editor.layout();
        delete this.layoutDebounce;
      }, DEBOUNCE_MSEC);
    }
  };

  public focusedEditor(): Editor | undefined {
    return [...this.editors.values()].find((editor) => editor.hasTextFocus());
  }

  public updateOptions(options: MonacoType.editor.IEditorOptions) {
    for (const editor of this.editors.values()) editor.updateOptions(options);
  }

  //=== Listen for user edits

  private ignoreAllEdits() {
    for (const editor of this.editors.values()) this.ignoreEdits(editor);
  }

  private ignoreEdits(editor: Editor) {
    editor.onDidChangeModelContent(() => {
      // no-op
    });
  }

  private observeAllEdits() {
    for (const editor of this.editors.values()) this.observeEdits(editor);
  }

  private observeEdits(editor: Editor) {
    const disposable = editor.onDidChangeModelContent(() => {
      this.isEdited ||= true;
      disposable.dispose();
    });
  }
}
