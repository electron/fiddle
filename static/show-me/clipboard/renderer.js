window.addEventListener('DOMContentLoaded', () => {
  const copyButton = document.querySelector('#copy')
  const pasteButton = document.querySelector('#paste')
  const textarea = document.querySelector('textarea')
  copyButton.onclick = () => {
    window.clipboard.writeText('Hello from Electron!')
  }
  pasteButton.onclick = async () => {
    textarea.value = await window.clipboard.readText()
  }
})
