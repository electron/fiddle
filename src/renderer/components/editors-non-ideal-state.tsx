import * as React from 'react';

import { Button, NonIdealState } from '@blueprintjs/core';

import { EditorMosaic } from '../editor-mosaic';

export function renderNonIdealState(editorMosaic: EditorMosaic) {
  const resolveButton = (
    <Button text="Reset editors" onClick={() => editorMosaic.resetLayout()} />
  );

  return (
    <NonIdealState
      action={resolveButton}
      icon="applications"
      description="You have closed all editors. You can open them again with the button below or in the sidebar menu!"
    />
  );
}
