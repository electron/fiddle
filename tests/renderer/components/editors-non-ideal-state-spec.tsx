import { mount } from 'enzyme';

import { renderNonIdealState } from '../../../src/renderer/components/editors-non-ideal-state';
import { EditorMosaic } from '../../../src/renderer/editor-mosaic';

describe('renderNonIdealState()', () => {
  let editorMosaic: EditorMosaic;

  beforeEach(() => {
    ({ editorMosaic } = (window as any).ElectronFiddle.app.state);
  });

  it('renders a non-ideal state', () => {
    expect(renderNonIdealState({} as any)).toMatchSnapshot();
  });

  it('handles a click', () => {
    const resetLayoutSpy = jest.spyOn(editorMosaic, 'resetLayout');
    const wrapper = mount(renderNonIdealState(editorMosaic as any));
    wrapper.find('button').simulate('click');
    expect(resetLayoutSpy).toHaveBeenCalledTimes(1);
  });
});
