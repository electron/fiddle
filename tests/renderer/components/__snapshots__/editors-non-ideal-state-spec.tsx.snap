// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderNonIdealState() renders a non-ideal state 1`] = `
<Blueprint3.NonIdealState
  action={
    <Blueprint3.Button
      onClick={[Function]}
      text="Reset editors"
    />
  }
  description="You have closed all editors. You can open them again with the button below or in the sidebar menu!"
  icon="applications"
/>
`;
